examples/r-future/r-furrr.Rmd

```diff
@@ -0,0 +1,210 @@
+
+> **Note**: We disable XGBoost's multi-threading in this example to show the difference between furrr and purrr for single-threaded workloads.
+
+```{r train model}
+train_model <- function(params, dtrain) {
+    model <- xgb.train(
+        params = params,
+        data = dtrain,
+        nrounds = 100,
+        nthread = 1,
+        objective = "reg:squarederror",
+    )
+    return(model)
+}
+
+test_results <- function(model, dtest) {
+    results <- predict(model, dtest)
+    return(results)
+}
+
+create_model_table <- function(data, ...) {
+    dmatrices <- create_matrices(data)
+
+    dtrain <- dmatrices$train
+    dtest <- dmatrices$test
+
+    params <- list(...)
+
+    model <- train_model(params, dtrain)
+    results <- test_results(model, dtest)
+    return(
+        tibble(
+            mean_absolute_error = mae(getinfo(dtest, "label"), results),
+            params = params
+        )
+    )
+}
+```
+
+### Let's Run the Model!
+It's finally time to run the functions we just created. We also define the hyperparameters we want to test. We use the expand_grid function (with `list()` around constants) to create the appropriate. Here we are only testing max_depth, but other hyperparameters could be added to the grid.
+
+Here are all the preprocessing steps.
+
+```{r preprocess}
+births_raw_data <- download_data()
+births_data <- filter_data(births_raw_data)
+births_data_preprocessed <- preprocess_data(births_data)
+
+params <- expand_grid(
+    data = list(births_data_preprocessed),
+    max_depth = seq(1, 8)
+)
+```
+
+#### Try Running the Model with purrr
+Let's first try it with purrr to iterate through the various models.
+
+Actually, let's not -- it takes over 17 minutes to run this code.  
+```{r purrr evaluation, eval = FALSE}
+
+# ###################################################
+# ## Don't actually run this - it takes forever... ##
+# ###################################################
+
+# print("With purrr:")
+# tic()
+# results_purrr <- pmap_dfr(params, create_model_table)
+# toc()
+
+# best_result_purrr <- results_purrr %>%
+#     top_n(-1, mean_absolute_error) %>%
+#     head(1)
+
+# print(best_result_purrr)
+```
+
+#### Run the Model with furrr
+
+Instead, let's try it with a parallel back-end. In this case, we ask for eight workers for eight processes.
+
+The only change to the purrr code is to add the futures `plan` and exchanging `pmap_dfr` with `future_pmap_dfr`. 
+
+```{r furrr evaluation}
+#####################
+## Run this instead##
+#####################
+plan(multisession(workers = 8))
+
+print("With furrr:")
+tic()
+results_furrr <- future_pmap_dfr(params, create_model_table, .options = furrr_options(seed = NULL))
+toc()
+
+best_result_furrr <- results_furrr %>%
+    top_n(-1, mean_absolute_error) %>%
+    head(1)
+
+print(best_result_furrr)
+
+plan(sequential)
```

Comment:
```suggestion
```

I actually think it's a best practice to not change it back to sequential after you're done. In a sense if you want sequential you just would use purrr instead of furrr